content/general/concepts/abstract-factory/abstract-factory.md

```diff
@@ -0,0 +1,381 @@
+                                   "Potter",
+                                   LocalDate.of(1980, 7, 31),
+                                   "4 Privet Drive, Little Whinging",
+                                   795);
+      case "Ron" -> new Customer("Ron",
+                                 "Weasley",
+                                 LocalDate.of(1980, 3, 1),
+                                 "Shell Cottage, Cornwall",
+                                 379);
+      case "Hermione" -> new Customer("Hermione",
+                                      "Granger",
+                                      LocalDate.of(1979, 9, 19),
+                                      "12 Grimmauld Palace, London",
+                                      843);
+      case "Lucius" -> new Customer("Lucius",
+                                    "Malfoy",
+                                    LocalDate.of(1954, 2, 4),
+                                    "Malfoy Mannor, Wiltshire",
+                                    978);
+      case "Dobby" -> new Customer("Dobby",
+                                   "the house elf",
+                                   LocalDate.of(1972, 6, 28),
+                                   "Malfoy Mannor, Wiltshire",
+                                   256);
+      case "Sirius" -> new Customer("Sirius",
+                                    "Black",
+                                    LocalDate.of(1959, 11, 3),
+                                    "12, Grimmauld Place, London",
+                                    169);
+      default -> throw new IllegalArgumentException("Could not return the credit history for " + name);
+    };
+  }
+}
+```
+
+The `CreditAgencyGateway` uses a switch statement to query a provided `name`. If the name is recognised a new customer is returned, else an exception is thrown.
+
+One of the main advantages to factory patterns, are they allow for a large amount of model classes and enforce a common interface between them. Below provides an abstract model `CurrentAccount` class for its concrete sub-classes to be based on:
+
+```java
+public abstract class CurrentAccount {
+  private final String accountUID = UUID.randomUUID().toString();
+  private final Customer accountHolder;
+  private int balance;
+  private int agreedOverdraftLimit;
+
+  public CurrentAccount(Customer accountHolder, int initBalance, int agreedOverdraftLimit) {
+    this.accountHolder = accountHolder;
+    this.balance = initBalance;
+    this.agreedOverdraftLimit = agreedOverdraftLimit;
+  }
+
+  public void addFunds(final int amount) {
+    balance += amount;
+  }
+
+  public void withdrawFunds(final int amount) {
+    if (amount < balance + agreedOverdraftLimit) {
+      balance -= amount;
+    } else {
+      System.out.println("Insufficient funds. Unable to withdraw " + amount);
+    }
+  }
+  // Enforces concrete classes to override this method
+  public abstract void increaseOverdraft(int requestedLimit);
+
+  // Getters
```

Comment:
Was there supposed to be more code here? Like getters and setters?

Reply:
The comment "// Getters" is placed there to let people know what additional boiler plate code is needed in order to make this work. The issue with writing in the getters, setters, toString etc is it massively increases the code e.g. without adding any new knowledge. This kind of repetitive non-business code is so common in java that a lot of industry developers use something called project Lombok (which will add code for should u annotate the class with @Getter). You can find out more (if your interested)  https://www.baeldung.com/intro-to-project-lombok

Reply:
Yes getters and setter might make sence here. Both ballance and agreedOverdraft are not final. These properties do however have acting setters programmed out. withdrawFunds() and addFunds() are essentially setters for balance, and increaseOverdraft is a setter for agreedOverdraft. 

Adding additional setters would cause a security concern as the method would be public (anything can access it) and someone could write...

CurrentAcount myAccount = new GoldPersonal(customer, int, int);
myAccount.setBalance(1_000_000);

That way they would get around the addFunds checks :)